constraints/miden-vm/stack.air

```diff
@@ -0,0 +1,957 @@
+        f_caller(b),
+        f_movup2(b),
+        f_movdn2(b),
+        f_movup3(b),
+        f_movdn3(b),
+        f_advpopw(b),
+        f_expacc(b),
+        f_movup4(b),
+        f_movdn4(b),
+        f_movup5(b),
+        f_movdn5(b),
+        f_movup6(b),
+        f_movdn6(b),
+        f_movup7(b),
+        f_movdn7(b),
+        f_swapw(b),
+        f_ext2mul(b),
+        f_movup8(b),
+        f_movdn8(b),
+        f_swapw2(b),
+        f_swapw3(b),
+        f_swapdw(b),
+        f_001_1111(b),
+
+        f_assert(b),
+        f_eq(b),
+        f_add(b),
+        f_mul(b),
+        f_and(b),
+        f_or(b),
+        f_u32and(b),
+        f_u32xor(b),
+        f_frie2f4(b),
+        f_drop(b),
+        f_cswap(b),
+        f_cswapw(b),
+        f_mloadw(b),
+        f_mstore(b),
+        f_mstorew(b),
+        f_fmpupdate(b),
+
+        f_pad(b),
+        f_dup(b),
+        f_dup1(b),
+        f_dup2(b),
+        f_dup3(b),
+        f_dup4(b),
+        f_dup5(b),
+        f_dup6(b),
+        f_dup7(b),
+        f_dup9(b),
+        f_dup11(b),
+        f_dup13(b),
+        f_dup15(b),
+        f_advpop(b),
+        f_sdepth(b),
+        f_clk(b),
+
+        f_u32add(b),
+        f_u32sub(b),
+        f_u32mul(b),
+        f_u32div(b),
+        f_u32split(b),
+        f_u32assert2(b),
+        f_u32add3(b),
+        f_madd(b),
+
+        f_hperm(b),
+        f_mpverify(b),
+        f_pipe(b),
+        f_mstream(b),
+        f_span(b),
+        f_join(b),
+        f_split(b),
+        f_loop(b),
+
+        f_mrupdate(b, extra),
+        f_push(b, extra),
+        f_syscall(b, extra),
+        f_call(b, extra),
+        f_end(b, extra),
+        f_repeat(b, extra),
+        f_respan(b, extra),
+        f_halt(b, extra)
+    ]
+
+
+### Helper evaluators #############################################################################
+
+# Enforces that the provided columns must be zero.
+ev is_zero(main: [values[4]]):
+    enf v = 0 for v in values
+
+
+# Enforces that values in the columns in the current row must be equal to the values in the next 
+# row.
+ev is_unchanged(main: [values[4]]):
+    enf v' = v for v in values
+
+
+# Enforces that the provided columns must be binary.
+ev is_binary(main: [a]):
+    enf a^2 = a
+
+
+ev range_check_16bit(main: [helper[6]], aux: [b_range]):
+    enf b_range' * ($alpha[0] + helper[0]) * ($alpha[1] * helper[1]) * ($alpha[2] + helper[2]) * ($alpha[3] * helper[3]) = b_range
+
+
+ev check_element_validity(main: [helper[6]]):
+    let m = helper[4]
+    let v_hi = 2^16 * helper[3] + helper[2]
+    let v_lo = 2^16 * helper[1] + helper[0]
+    enf (1 - m * (2^32 - 1 - v_hi)) * v_lo = 0
+
+
+### Stack Air Constraints #########################################################################
+
+# Enforces the constraints on the stack.
+# TODO: add docs for columns
+ev stack_constraints(main: [stack[16], bookkeeping[2], h0, helper[6], clk, fmp], aux: [op_bits[7], extra, hasher5, b_chip, b_range]):
```

Comment:
A few comments on these:

- I would probably change `stack` to `stack_top`.
- I'd combine `bookkeeping` and `h0` into something like `stack_helpers`.
- I'd rename `helpers` to `op_helpers`.
- `op_bits`, `extra` are part of the main trace - not auxiliary trace. Also, I'd probably lump `extra` together with `op_bits`.
- What is `hasher5`?

Overall, this could loo like so:
```
ev stack_constraints(main: [stack_top[16], stack_helpers[2], op_helper[6], op_bits[8], clk, fmp], aux: [hasher5, b_chip, b_range]):
```